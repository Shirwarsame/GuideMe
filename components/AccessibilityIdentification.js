import React from 'react';
import { Text, View, Image } from 'react-native';
import { Marker } from 'react-native-maps';
import { styles } from '../screens/Map'


function AccessibilityIdentification() {
    let handicapedSignSource = './../assets/Icons/handicap_icon.png';

        return(
            <View>
                    <Marker
                    coordinate={{ latitude: 45.497245, longitude: -73.578968}}>
                    <Image style={styles.AccessibilityIdentification} source={require(handicapedSignSource)}/>
                    </Marker>

                    <Marker
                    coordinate={{ latitude: 45.496612, longitude: -73.578010}}>
                    <Image style={styles.AccessibilityIdentification} source={require(handicapedSignSource)}/>
                    </Marker>

                    <Marker
                    coordinate={{ latitude: 45.495899, longitude:  -73.578712}}>
                    <Image style={styles.AccessibilityIdentification} source={require(handicapedSignSource)}/>
                    </Marker>
            </View>
            );
    }


export {AccessibilityIdentification};