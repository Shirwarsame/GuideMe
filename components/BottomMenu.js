/* eslint-disable no-unused-vars */
import React, { useLayoutEffect, useEffect } from "react";
import { View, AsyncStorage, Text, StyleSheet, Switch } from "react-native";
import { Icon } from "native-base";
import { Button } from "react-native-paper";
import { store } from "../redux/reducers/index";
import { FloorMenu } from "./FloorMenu";
import {CampusRegion} from "../constants/buildingData";


/**
 * US6 - As a user, I would like to switch between the SGW and the Loyola maps
 * The following function renders a menu at the bottom of the screen. The menu
 * includes a toggle (US6) & an arrow icon leading to the More Details page.
 */

function BottomMenu (props) {
    const [selectedBuilding, setSelectedBuilding] = React.useState(null);
    const [swtichCampus, setSwitchCampus] = React.useState(true);
    const [destination, setDestination] = React.useState("");
    const [methodTravel, setMethodTravel] = React.useState("driving");
    const [personaType, setPersonaType] = React.useState("");
    const [mobilityReduced, setMobilityReduced] = React.useState("");
    const viewIndoor = props.indoor;
    const inDirections = props.inDirections;
    const building = props.building;
    const previewDirections = props.previewMode;
    const from = props.from;
    const to = props.to;
    

    const goBack = () => {
        props.navigation.goBack();
    };

    const goToDoubleSearchBar = () => {
        props.navigation.navigate("DoubleSearch", { destinationName: destination });
    };
    const goToDirections = () => {
        props.navigation.navigate("Directions", { destinationResponse: props.directionResponse });
    };

    const goToPreferenceMenu = (isIndoor) => {
        if (isIndoor) {
            props.navigation.navigate("PreferenceMenu", {
                indoor: true,
                personaType: personaType,
                mobilityType: mobilityReduced,
                transportType: methodTravel
            });
        } else {
            props.navigation.navigate("PreferenceMenu", {
                indoor: false,
                personaType: personaType,
                mobilityType: mobilityReduced,
                transportType: methodTravel
            });
        }

    };

    const goToMoreDetails = () => {
        props.navigation.navigate("MoreDetails", {
            name: store.getState().selectedBuildingName
        });
    };

    const goToInsideBuilding = () => {
        props.navigation.navigate("IndoorMapView", { selectedBuilding: selectedBuilding });
    };

    const goToNearby = () => {
        AsyncStorage.setItem("sideMenu", "mapView"); 
        props.navigation.navigate("NearbyInterest", { campusBool: swtichCampus});
    
    };

    
    useLayoutEffect(() => {
        const unsubscribe = store.subscribe(() => {
            setSelectedBuilding(store.getState().selectedBuildingName);
            setDestination(store.getState().mainSearchBarDestination);
            setPersonaType(store.getState().personaType);
            setMobilityReduced(store.getState().mobilityReducedType);
            setMethodTravel(store.getState().transportType);
        });
        return function cleanUp() {
            unsubscribe();
        }
    });


    const switchCampuses = (value) => {
        setSwitchCampus(value);
        if (value) {
            props.mapReference.current.animateToRegion(CampusRegion.sgwCoord);        }
        else {
            props.mapReference.current.animateToRegion(CampusRegion.loyCoord);
        }
    }

    if (viewIndoor) {
        return (
            <View style={styles.insideBuildingContainer} data-test="BottomMenu" testID="bottomMenuInitalView">
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={() => { (inDirections ? goToPreferenceMenu(true) : goToMoreDetails) }} />
                <Text style={styles.mainLabel}>{building ? building : selectedBuilding}</Text>
                <Text style={styles.shortLabel}>More info</Text>
                <Button testID="indoorMapExitBuildingButton" style={styles.btnleave} color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goBack}>
                    <Text style={styles.btnText}>Exit Building</Text>
                </Button>
                <View style={styles.changeFloor}>
                    <FloorMenu from={from} to={to} />
                </View>
            </View>
        )
    }

    if (previewDirections) {
        return (
            <View style={styles.container}>
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={() => { goToPreferenceMenu(false) }} />

                {props.directionResponse && props.directionResponse.generalRouteInfo.totalDuration.length > 12 &&
                    <Text style={styles.mainLabel}>{props.directionResponse.generalRouteInfo.totalDuration}</Text>
                }
                 {props.directionResponse && props.directionResponse.generalRouteInfo.totalDuration.length <= 12 &&
                    <Text style={styles.mainLabel}>{props.directionResponse.generalRouteInfo.totalDuration } {props.directionResponse.generalRouteInfo.totalDistance}</Text>
                }
                 {!props.directionResponse &&
                    <Text style={styles.mainLabel}>N/A</Text>
                }
                <Text style={styles.shortLabel}>Main Travel Mode: {methodTravel}</Text>
                <View style={styles.btnGetDirection}>
                    <Button style={styles.btnGetDirectionPosition}
                        color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goToDirections}>
                        <Text style={styles.btnText}>Start</Text>
                    </Button>
                </View>
            </View>
        );
    }
    if (destination) {
        if (destination.length > 13) {
            var updatedDestination = destination.substring(0, 13) + "...";
        }
        return (
            <View style={styles.container} >
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={() => { props.navigation.navigate("MoreDetails", { name: destination }) }} />
                {destination.length > 13
                    ? <Text style={styles.mainLabel}>{updatedDestination}</Text>
                    : <Text style={styles.mainLabel}>{destination}</Text>
                }
                <Text style={styles.shortLabel}>More info</Text>
                <View style={styles.btnGetDirection}>
                    <Button testID="getDirectionsButton" style={styles.btnGetDirection} color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goToDoubleSearchBar}>
                        <Text style={styles.btnText}>Get Directions</Text>
                    </Button>
                </View>
            </View>
        );
    }

    if (selectedBuilding) {
        if (selectedBuilding.length > 13) {
            var updatedSelectedBuilding = selectedBuilding.substring(0, 13) + "...";

        }
        return (
            <View style={styles.container}>
                <Icon testID="getInsideIcon" name="ios-arrow-up" style={styles.arrowUp} onPress={goToMoreDetails} />
                <Text style={styles.mainLabel}>{selectedBuilding}</Text>
                {selectedBuilding.length > 13
                    ? <Text style={styles.mainLabel}>{updatedSelectedBuilding}</Text>
                    : <Text style={styles.mainLabel}>{selectedBuilding}</Text>
                }
                <Text style={styles.shortLabel}>More info</Text>
                <Button testID="getInsideButton" style={styles.btn} color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goToInsideBuilding}>
                    <Text style={styles.btnText}>Get Inside</Text>
                </Button>
            </View>
        );
    }
    else {
        return (
            <View style={styles.container} data-test="BottomMenu" testID="bottomMenuInitalView">
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={goToNearby} />
                <Text style={styles.mainLabel}>Nearby</Text>
                <Text style={styles.shortLabel}>Food, drinks & more</Text>
                <View testID="intialBottomMenuToggleButton" style={styles.toggle}>
                    <Switch
                        value={swtichCampus}
                        onValueChange={(val) => { switchCampuses(val);}}>
                    </Switch>
                </View>
            </View>
        );
    }
}

export const styles = StyleSheet.create({
    container: {
        width: "100%",
        height: 350,
        position: "absolute",
        borderRadius: 30.5,
        backgroundColor: "#2A2E43",
        bottom: -275
    },
    insideBuildingContainer: {
        width: "100%",
        height: 425,
        position: "absolute",
        borderRadius: 30.5,
        backgroundColor: "#2A2E43",
        bottom: -275
    },
    moreDetails: {
        width: "100%",
        height: "100%",
        position: "absolute",
        backgroundColor: "#2A2E43",
        alignItems: "center",
        justifyContent: "space-between"
    },
    arrowUp: {
        color: "#ffffff",
        left: "5%",
        top: "7%",
    },
    toggle: {
        position: "absolute",
        left: "80%",
        top: "7%"
    },
    btn: {
        position: "absolute",
        left: "65%",
        top: "5.5%",
        color: "#FFFFFF"
    },
    btnGetDirection: {
        position: "absolute",
        left: "57%",
        top: "5.5%",
        color: "#FFFFFF"
    },
    btnGetDirectionPosition: {
        left: "100%"
    },
    btnleave: {
        position: "absolute",
        left: "62%",
        top: "5.5%",
        color: "#FFFFFF",
    },
    btnText: {
        color: "#FFFFFF",
        fontFamily: "encodeSansExpanded"
    },
    mainLabel: {
        position: "absolute",
        top: "5%",
        left: "12.5%",
        color: "#FFFFFF",
        fontSize: 20,
        fontFamily: "encodeSansExpanded",
        height: "9%",
        width: "45%"
    },
    shortLabel: {
        position: "absolute",
        top: "12%",
        left: "12.5%",
        color: "#80828D",
        fontSize: 16,
        fontFamily: "encodeSansExpanded"
    },
    arrowDown: {
        color: "#ffffff",
        top: "5%",
        fontSize: 54,
        position: "absolute"
    },
    changeFloor: {
        top: "12.5%",
        left: "15%",
    }
});

export { BottomMenu };