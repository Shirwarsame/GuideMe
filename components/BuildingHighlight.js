import React from "react";
import { View } from "react-native";
import { Polygon } from "react-native-maps";
import coord from "../constants/buildingCoordinates";
import { buildingData } from "../constants/buildingData";
import { connect } from "react-redux";


function mapStateToProps(buildingName) {
    return {
        selectedBuildingName: buildingName
    }
}

function mapDispatchToProps(dispatch) {
    return {
        setSelectedBuildingName: (value) => dispatch({ type: "UPDATE_SELECTED_BUILDING", payload: value }),
    }
}

/**
 * US3 - As a user, I would like to be able to identify campus buildings and
 * distinguish them from other buildings.
 * The following function colors the campus buildings.
 */

export function BuildingHighlight (props) {
    const [buildingName, setBuildingName] = React.useState("");
    var buildings = buildingData();

    return (
        <View>
            <Polygon
                id="tap_h"
                coordinates={coord.h.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["Hall Building"].name); 
                    props.setSelectedBuildingName(buildings["Hall Building"].name);
            }}
                fillColor={buildingName == buildings["Hall Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_lb"
                coordinates={coord.lb.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["LB Building"].name);
                    props.setSelectedBuildingName(buildings["LB Building"].name);
                } }
                fillColor={buildingName == buildings["LB Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_gm"
                coordinates={coord.gm.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["GM Building"].name)
                    props.setSelectedBuildingName(buildings["GM Building"].name);
                }}
                fillColor={buildingName == buildings["GM Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_ev"
                coordinates={coord.ev.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["EV Building"].name)
                    props.setSelectedBuildingName(buildings["EV Building"].name);
                }}
                fillColor={buildingName == buildings["EV Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_mb"
                coordinates={coord.mb.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["MB Building"].name);
                    props.setSelectedBuildingName(buildings["MB Building"].name);
                }}
                fillColor={buildingName == buildings["MB Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_sp"
                coordinates={coord.sp.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["SP Building"].name);
                    props.setSelectedBuildingName(buildings["SP Building"].name);
                }}
                fillColor={buildingName == buildings["SP Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_cj"
                coordinates={coord.cj.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["CJ Building"].name)
                    props.setSelectedBuildingName(buildings["CJ Building"].name);
            }}
            fillColor={buildingName == buildings["CJ Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_cc"
                coordinates={coord.cc.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["CC Building"].name);
                    props.setSelectedBuildingName(buildings["CC Building"].name);
            }}
            fillColor={buildingName == buildings["CC Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_ad"
                coordinates={coord.ad.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["AD Building"].name);
                    props.setSelectedBuildingName(buildings["AD Building"].name);
                }}
                fillColor={buildingName == buildings["AD Building"].name ? "#3ACCE1": "#a282e0"}
                />

            <Polygon
                id="tap_gn"
                coordinates={coord.gn.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["Grey Nuns"].name)
                    props.setSelectedBuildingName(buildings["Grey Nuns"].name);
                }}
                fillColor={buildingName == buildings["Grey Nuns"].name ? "#3ACCE1": "#a282e0"}
            />
            <Polygon
                id="tap_vl"
                coordinates={coord.vl.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["VL Building"].name)
                    props.setSelectedBuildingName(buildings["VL Building"].name);
            }}
            fillColor={buildingName == buildings["VL Building"].name ? "#3ACCE1": "#a282e0"}
            />
        </View>
    );
}

export default connect(mapStateToProps, mapDispatchToProps)(BuildingHighlight);
