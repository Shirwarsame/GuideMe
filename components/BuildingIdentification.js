import React from 'react';
import { Text, View } from 'react-native';
import { styles } from '../screens/Map'
import { Marker } from 'react-native-maps';

function BuildingIdentification () {
    return (
        <View>
            <Marker
                coordinate={{ latitude: 45.497222, longitude: -73.578809 }}
                title={'H building'}>
                <Text style={styles.buildingIdentification}>H</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.496537, longitude: -73.577755 }}
                title={'LB building'}>
                <Text style={styles.buildingIdentification}>LB</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.495804, longitude: -73.578764 }}
                title={'HM building'}>
                <Text style={styles.buildingIdentification}>HM</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.495431, longitude: -73.578133 }}
                title={'EV building'}>
                <Text style={styles.buildingIdentification}>EV</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.495306, longitude: -73.579011 }}
                title={'JMSB building'}>
                <Text style={styles.buildingIdentification}>JMSB</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.457825, longitude: -73.641458 }}
                title={'SP'}>
                <Text style={styles.buildingIdentification}>SP</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.457516, longitude: -73.640327 }}
                title={'CJ'}>
                <Text style={styles.buildingIdentification}>CJ</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.458094, longitude: -73.639768 }}
                title={'AP'}>
                <Text style={styles.buildingIdentification}>AP</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.458314, longitude: -73.640409 }}
                title={'CC'}>
                <Text style={styles.buildingIdentification}>CC</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.493450, longitude: -73.576814 }}
                title={'GN'}>
                <Text style={styles.buildingIdentification}>GN</Text>
            </Marker>
        </View>
    );
}

export { BuildingIdentification };