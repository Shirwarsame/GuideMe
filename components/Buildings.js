import React from 'react';
import { View } from 'react-native';
import { Polygon } from 'react-native-maps';

const Buildings = () => {
    return(
        <View>
            <Polygon
                coordinates={[
                    { latitude: 45.497167, longitude: -73.579543 },
                    { latitude: 45.497713, longitude: -73.579032 },
                    { latitude: 45.497378, longitude: -73.578340 },
                    { latitude: 45.496830, longitude: -73.578853 },
                    { latitude: 45.497167, longitude: -73.579543 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
           />

            <Polygon
                coordinates={[
                    { latitude: 45.496708, longitude: -73.578525 },
                    { latitude: 45.497291, longitude: -73.578017 },
                    { latitude: 45.496896, longitude: -73.577102 },
                    { latitude: 45.496255, longitude: -73.577700 },
                    { latitude: 45.496708, longitude: -73.578525 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.495782, longitude: -73.579149 },
                    { latitude: 45.496129, longitude: -73.578806 },
                    { latitude: 45.495945, longitude: -73.578429 },
                    { latitude: 45.495616, longitude: -73.578742 },
                    { latitude: 45.495782, longitude: -73.579149 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.495574, longitude: -73.578653 },
                    { latitude: 45.495908, longitude: -73.578445 },
                    { latitude: 45.495416, longitude: -73.577551 },
                    { latitude: 45.495172, longitude: -73.577854 },
                    { latitude: 45.495574, longitude: -73.578653 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.495523, longitude: -73.579198 },
                    { latitude: 45.495358, longitude: -73.579356 },
                    { latitude: 45.495014, longitude: -73.578734 },
                    { latitude: 45.495200, longitude: -73.578530 },
                    { latitude: 45.495523, longitude: -73.579198 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.457464, longitude: -73.642074 },
                    { latitude: 45.458325, longitude: -73.641411 },
                    { latitude: 45.458165, longitude: -73.640985 },
                    { latitude: 45.457528, longitude: -73.641481 },
                    { latitude: 45.457198, longitude: -73.640661 },
                    { latitude: 45.457002, longitude: -73.640830 },
                    { latitude: 45.457464, longitude: -73.642074 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.457335, longitude: -73.640715 },
                    { latitude: 45.457179, longitude: -73.640400 },
                    { latitude: 45.457617, longitude: -73.640051 },
                    { latitude: 45.457826, longitude: -73.640483 },
                    { latitude: 45.457335, longitude: -73.640715 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.458521, longitude: -73.640679 },
                    { latitude: 45.458377, longitude: -73.640794 },
                    { latitude: 45.458076, longitude: -73.639997 },
                    { latitude: 45.458221, longitude: -73.639898 },
                    { latitude: 45.458521, longitude: -73.640679}
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.457901, longitude: -73.640113 },
                    { latitude: 45.458378, longitude: -73.639742 },
                    { latitude: 45.458243, longitude: -73.639460 },
                    { latitude: 45.457807, longitude: -73.639814 },
                    { latitude: 45.457901, longitude: -73.640113 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />
            
        </View>
    )
}

export { Buildings };
