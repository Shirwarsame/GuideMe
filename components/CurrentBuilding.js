import React, { useEffect, useState } from 'react';
import coord from '../constants/buildingCoordinates';
import { isPointInPolygon } from 'geolib'

export function CurrentBuilding () {
    const [currentBuilding, setcurrentBuilding] = React.useState("")
    const [lastLat, setlastLat] = React.useState(0)
    const [lastLong, setlastLong] = React.useState(0)
    const [altitude, setAltitude] = React.useState(0) //May not work since it is a number (we may need a string) 

    AsyncStorage.setItem("altitude", altitude);

    useEffect(() => {
            this.watchID = navigator.geolocation.watchPosition((position) => {
                setlastLat(position.coords.latitude);
                setlastLong(position.coords.longitude);
                setAltitude(position.coords.altitude);
            },
                 (error) => alert(error.message),
                 { enableHighAccuracy: true, timeout: 20000, maximumAge: 1000 }
            );
        const intervalId = setInterval(() => {
            if (isPointInPolygon({ latitude: lastLat, longitude: lastLong },
                [{ latitude: coord.gn.coordinates[0].latitude, longitude: coord.gn.coordinates[0].longitude },
                { latitude: coord.gn.coordinates[1].latitude, longitude: coord.gn.coordinates[1].longitude },
                { latitude: coord.gn.coordinates[2].latitude, longitude: coord.gn.coordinates[2].longitude },
                { latitude: coord.gn.coordinates[3].latitude, longitude: coord.gn.coordinates[3].longitude },
                { latitude: coord.gn.coordinates[4].latitude, longitude: coord.gn.coordinates[4].longitude }])) {
    
                setcurrentBuilding(coord.gn.name)
    
            }
    
            if (isPointInPolygon({ latitude: lastLat, longitude: lastLong },
                [{ latitude: coord.h.coordinates[0].latitude, longitude: coord.h.coordinates[0].longitude },
                { latitude: coord.h.coordinates[1].latitude, longitude: coord.h.coordinates[1].longitude },
                { latitude: coord.h.coordinates[2].latitude, longitude: coord.h.coordinates[2].longitude },
                { latitude: coord.h.coordinates[3].latitude, longitude: coord.h.coordinates[3].longitude },
                { latitude: coord.h.coordinates[4].latitude, longitude: coord.h.coordinates[4].longitude }])) {
    
                setcurrentBuilding(coord.h.name)
    
            }
    
            if (isPointInPolygon({ latitude: lastLat, longitude: lastLong },
                [{ latitude: coord.mb.coordinates[0].latitude, longitude: coord.mb.coordinates[0].longitude },
                { latitude: coord.mb.coordinates[1].latitude, longitude: coord.mb.coordinates[1].longitude },
                { latitude: coord.mb.coordinates[2].latitude, longitude: coord.mb.coordinates[2].longitude },
                { latitude: coord.mb.coordinates[3].latitude, longitude: coord.mb.coordinates[3].longitude },
                { latitude: coord.mb.coordinates[4].latitude, longitude: coord.mb.coordinates[4].longitude }])) {
    
                setcurrentBuilding(coord.mb.name)
    
            }
    
            if (isPointInPolygon({ latitude: lastLat, longitude: lastLong },
                [{ latitude: coord.ev.coordinates[0].latitude, longitude: coord.ev.coordinates[0].longitude },
                { latitude: coord.ev.coordinates[1].latitude, longitude: coord.ev.coordinates[1].longitude },
                { latitude: coord.ev.coordinates[2].latitude, longitude: coord.ev.coordinates[2].longitude },
                { latitude: coord.ev.coordinates[3].latitude, longitude: coord.ev.coordinates[3].longitude },
                { latitude: coord.ev.coordinates[4].latitude, longitude: coord.ev.coordinates[4].longitude }])) {
    
                setcurrentBuilding(coord.ev.name)
    
            }
    
            if (isPointInPolygon({ latitude: lastLat, longitude: lastLong },
                [{ latitude: coord.lb.coordinates[0].latitude, longitude: coord.lb.coordinates[0].longitude },
                { latitude: coord.lb.coordinates[1].latitude, longitude: coord.lb.coordinates[1].longitude },
                { latitude: coord.lb.coordinates[2].latitude, longitude: coord.lb.coordinates[2].longitude },
                { latitude: coord.lb.coordinates[3].latitude, longitude: coord.lb.coordinates[3].longitude },
                { latitude: coord.lb.coordinates[4].latitude, longitude: coord.lb.coordinates[4].longitude }])) {
    
                setcurrentBuilding(coord.lb.name)
    
            }
        }, 10000)
        return () => clearInterval(intervalId);
    })

    if (currentBuilding === "") {
        return ("Not in a building");
    }
    return (currentBuilding);
}