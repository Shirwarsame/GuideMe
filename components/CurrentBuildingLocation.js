import React, { useEffect } from "react";
import coord from "../constants/buildingCoordinates";
import { isPointInPolygon } from "geolib";
import { StyleSheet, TouchableOpacity  } from "react-native";
import MaterialIcons from "react-native-vector-icons/MaterialCommunityIcons";
import { connect } from "react-redux";

function mapStateToProps(state) {
    return {
        selectedBuildingName: state.selectedBuildingName
    }
}

function mapDispatchToProps(dispatch) {
    return {
        setCurrentBuilding: (value) => dispatch({ type: "UPDATE_SELECTED_BUILDING", payload: value }),
    }
}

/**
 * US5 - As a user, I would like to know which building I am currently in
 * US32 - As a user, I would like to be able to know where I am indoors.
 * The following function provides the building & current floor of a user.
 * 
 * Note: call CurrentBuildingLocation() inside BottomMenu.js
 */
export function CurrentBuildingLocation (props) {

    const [currentLocation, setCurrentLocation] = React.useState(null);


    var getPosition = function (options) {
        return new Promise(function (resolve, reject) {
            navigator.geolocation.getCurrentPosition(resolve, reject, options);
        });
    }

    const GetCurrentLocation = () => {

        getPosition().then(({ coords }) => {
            setCurrentLocation({
                latitude: coords.latitude,
                longitude: coords.longitude,
                altitude: coords.altitude
            })
        })
            .catch((err) => {
                alert(err.message);
            });
    }

    const checkIfWithinPolygon = () => {

        let withinBuilding = false;
        for (var key in coord) {
            if (isPointInPolygon(currentLocation, coord[key].coordinates)){
                withinBuilding = true;
                props.setCurrentBuilding(coord[key].name);
                break;
            }
        }
        if (withinBuilding) {
            props.mapReference.current.animateToRegion({
                latitude: dummyCurrentLocation.latitude,
                longitude: dummyCurrentLocation.longitude,
                latitudeDelta: 0.003,
                longitudeDelta: 0.003
              })
        }
        else {
        alert("You have to be in a highlighted concordia building to use this feature.");
        }  
    }

    //TODO: Uncomment this if you want to see how it works (This is GM Building)
    const dummyCurrentLocation = {
        altitude: 29.498906132077124,
        latitude: 45.495983,
        longitude: -73.578824
    }

    const onPressHandler = () => {
        if(currentLocation) {
            checkIfWithinPolygon();
        }
    }
    

    useEffect(() => {
        GetCurrentLocation();
    }, []);


    return (
        <TouchableOpacity style ={styles.layout} onPress={onPressHandler}>
            <MaterialIcons name="office-building" size={35} color="white"></MaterialIcons>
        </TouchableOpacity>
    );
}
export const styles = StyleSheet.create({
    layout: {
        width: 60,
        height: 60,
        borderRadius: 100/2,
        backgroundColor: "#2A2E43",
        justifyContent: "center",
        alignItems: "center"       
    },
    modal: {
        position:"absolute",
        right: "20%",
        paddingTop:10,
        borderRadius:10,
        width:200,
        backgroundColor: "white",
    },
    modalText: {
        fontSize:20,
        textAlign: 'center'      
    },
});

export default connect(mapStateToProps, mapDispatchToProps)(CurrentBuildingLocation);
