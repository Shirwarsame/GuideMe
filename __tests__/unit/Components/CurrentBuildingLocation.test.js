import React from "react";
import { CurrentBuildingLocation } from "../../../components/CurrentBuildingLocation";
import renderer from "react-test-renderer";
import { TouchableOpacity } from "react-native";
import { shallow } from 'enzyme';

describe("CurrentBuildingLocation component", () => {
    test("renders correctly", () => {
        const tree = renderer.create(<CurrentBuildingLocation />).toJSON();
        expect(tree).toMatchSnapshot();
      });

      test("click on navigate", () => {
        const getPosition = {
          checkIfWithinPoly
          }
        const checkIfWithinPoly ={
            getCurrentPosition: (coords) => {
                return coords;
              },
        }
        const wrapper = shallow(<CurrentBuildingLocation getPosition={getPosition}/>);
  
        wrapper.find(TouchableOpacity).at(0).simulate('onPress')
      });


});
