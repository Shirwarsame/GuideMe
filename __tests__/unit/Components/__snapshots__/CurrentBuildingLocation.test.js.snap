// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentBuildingLocation component renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#2A2E43",
      "borderRadius": 50,
      "height": 60,
      "justifyContent": "center",
      "opacity": 1,
      "width": 60,
    }
  }
>
  <Text />
</View>
`;
